src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Completion/RemoteCompletionService.cs

```diff
@@ -0,0 +1,229 @@
+        VSInternalClientCapabilities clientCapabilities,
+        RazorCompletionOptions razorCompletionOptions,
+        HashSet<string> existingHtmlCompletions,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetCompletionAsync(
+                context,
+                positionInfo,
+                completionContext,
+                clientCapabilities,
+                razorCompletionOptions,
+                existingHtmlCompletions,
+                cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<Response> GetCompletionAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        CompletionPositionInfo positionInfo,
+        VSInternalCompletionContext completionContext,
+        VSInternalClientCapabilities clientCapabilities,
+        RazorCompletionOptions razorCompletionOptions,
+        HashSet<string> existingDelegatedCompletions,
+        CancellationToken cancellationToken)
+    {
+        VSInternalCompletionList? csharpCompletionList = null;
+        var documentPositionInfo = positionInfo.DocumentPositionInfo;
+        if (documentPositionInfo.LanguageKind == RazorLanguageKind.CSharp &&
+            CompletionTriggerCharacters.IsValidTrigger(CompletionTriggerCharacters.CSharpTriggerCharacters, completionContext))
+        {
+            var mappedPosition = documentPositionInfo.Position;
+            csharpCompletionList = await GetCSharpCompletionAsync(
+                    remoteDocumentContext,
+                    documentPositionInfo.HostDocumentIndex,
+                    mappedPosition,
+                    positionInfo.ProvisionalTextEdit,
+                    completionContext,
+                    clientCapabilities,
+                    razorCompletionOptions,
+                    cancellationToken);
+
+            if (csharpCompletionList is not null)
+            {
+                Debug.Assert(existingDelegatedCompletions.Count == 0, "Delegated completion should be either C# or HTML, not both");
+                existingDelegatedCompletions.UnionWith(csharpCompletionList.Items.Select((item) => item.Label));
+            }
+        }
+
+        var razorCompletionList = CompletionTriggerCharacters.IsValidTrigger(CompletionTriggerCharacters.RazorTriggerCharacters, completionContext)
+            ? await _razorCompletionListProvider.GetCompletionListAsync(
+                documentPositionInfo.HostDocumentIndex,
+                completionContext,
+                remoteDocumentContext,
+                clientCapabilities,
+                existingCompletions: existingDelegatedCompletions,
+                razorCompletionOptions,
+                cancellationToken)
+            : null;
+
+        if (CompletionListMerger.Merge(razorCompletionList, csharpCompletionList) is not { } mergedCompletionList)
+        {
+            return Response.CallHtml;
```

Comment:
I see how it's confusing the way it's written, especially without any comments. Merge will return null only if both inputs are null, so essentially null merged list means we didn't get anything OOP and should just proceed with HTML. I see that it's not at all obvious without looking at the code for the Merge :). It is how the existing code is written, and I was essentially following the flow of the existing code as much as I can to prevent regressions. 

We also can't just ```return Merge(...)``` - if we didn't get anything from Razor or C# completion for some reason, Response.Result would be empty, and we'd still have to say CallHtml one more time. 

I'm adding a comment for now and would prefer to leave the logic as is, unless you strongly disagree.

Reply:
I think nobody has more context on this code than you, right now. I think cohosting is a great opportunity for us to improve our existing code and make things better. If you think the best way to take advantage of those things is more comments, then great. I love comments. If you think it can be written in a better way, then great.

Reply:
I'll leave it as is for now, hopefully the comment makes the intent here clear.